File: dream/scripts/verify_data.py
import json
import os
import sys

def load_json(file_path):
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            return json.load(f)
    except FileNotFoundError:
        print(f"Error: File not found at {file_path}")
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Failed to decode JSON from {file_path}: {e}")
        sys.exit(1)

def verify_data():
    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    pets_path = os.path.join(base_dir, 'data', 'soul_pets.json')
    skills_path = os.path.join(base_dir, 'data', 'skills.json')

    print(f"Loading data from:\n  {pets_path}\n  {skills_path}\n")

    pets = load_json(pets_path)
    skills = load_json(skills_path)

    # 1. Verify Skill IDs
    skill_map = {s['id']: s for s in skills}
    print(f"Loaded {len(skills)} skills.")

    skill_errors = 0
    for pet in pets:
        if 'skills' in pet:
            for pet_skill in pet['skills']:
                skill_id = pet_skill.get('skill_id')
                if not skill_id:
                    print(f"Warning: Pet {pet['name']} ({pet['id']}) has a skill without an ID.")
                    continue
                
                if skill_id not in skill_map:
                    print(f"Error: Pet {pet['name']} ({pet['id']}) references unknown skill ID: {skill_id}")
                    skill_errors += 1
                else:
                    # Optional: Check if name matches (just a warning if not)
                    db_name = skill_map[skill_id]['name']
                    if pet_skill.get('name') != db_name:
                        print(f"Note: Pet {pet['name']} calls skill {skill_id} '{pet_skill.get('name')}', but DB name is '{db_name}'.")

    # 2. Verify Evolution Target IDs
    pet_map = {p['id']: p for p in pets}
    print(f"Loaded {len(pets)} soul pets.")

    evo_errors = 0
    for pet in pets:
        if 'evolutions' in pet:
            for evo in pet['evolutions']:
                target_id = evo.get('target_id')
                if target_id not in pet_map:
                    print(f"Error: Pet {pet['name']} ({pet['id']}) evolves into unknown ID: {target_id}")
                    evo_errors += 1

    print("\n--- Verification Summary ---")
    if skill_errors == 0 and evo_errors == 0:
        print("✅ Data verification PASSED! All references are valid.")
    else:
        print(f"❌ Data verification FAILED with {skill_errors} skill errors and {evo_errors} evolution errors.")

if __name__ == "__main__":
    verify_data()

File: dream/scripts/verify_data.js
const fs = require('fs');
const path = require('path');

function loadJson(filePath) {
    try {
        const data = fs.readFileSync(filePath, 'utf8');
        return JSON.parse(data);
    } catch (err) {
        console.error(`Error loading ${filePath}: ${err.message}`);
        process.exit(1);
    }
}

function verifyData() {
    const baseDir = path.dirname(__dirname); // parent of scripts/ is dream/
    const petsPath = path.join(baseDir, 'data', 'soul_pets.json');
    const skillsPath = path.join(baseDir, 'data', 'skills.json');

    console.log(`Loading data from:\n  ${petsPath}\n  ${skillsPath}\n`);

    const pets = loadJson(petsPath);
    const skills = loadJson(skillsPath);

    // 1. Verify Skill IDs
    const skillMap = new Map(skills.map(s => [s.id, s]));
    console.log(`Loaded ${skills.length} skills.`);

    let skillErrors = 0;
    
    pets.forEach(pet => {
        if (pet.skills) {
            pet.skills.forEach(petSkill => {
                const skillId = petSkill.skill_id;
                if (!skillId) {
                    console.warn(`Warning: Pet ${pet.name} (${pet.id}) has a skill without an ID.`);
                    return;
                }

                if (!skillMap.has(skillId)) {
                    console.error(`Error: Pet ${pet.name} (${pet.id}) references unknown skill ID: ${skillId}`);
                    skillErrors++;
                } else {
                    const dbName = skillMap.get(skillId).name;
                    if (petSkill.name && petSkill.name !== dbName) {
                        console.log(`Note: Pet ${pet.name} calls skill ${skillId} '${petSkill.name}', but DB name is '${dbName}'.`);
                    }
                }
            });
        }
    });

    // 2. Verify Evolution Target IDs
    const petMap = new Map(pets.map(p => [p.id, p]));
    console.log(`Loaded ${pets.length} soul pets.`);

    let evoErrors = 0;
    pets.forEach(pet => {
        if (pet.evolutions) {
            pet.evolutions.forEach(evo => {
                const targetId = evo.target_id;
                if (!targetId || !petMap.has(targetId)) {
                    console.error(`Error: Pet ${pet.name} (${pet.id}) evolves into unknown ID: ${targetId}`);
                    evoErrors++;
                }
            });
        }
    });

    console.log("\n--- Verification Summary ---");
    if (skillErrors === 0 && evoErrors === 0) {
        console.log("✅ Data verification PASSED! All references are valid.");
    } else {
        console.error(`❌ Data verification FAILED with ${skillErrors} skill errors and ${evoErrors} evolution errors.`);
        process.exit(1);
    }
}

verifyData();
